File: WalletKitCore/src/walletkit/WKUnit.c
//
//  WKUnit.c
//  WalletKitCore
//
//  Created by Ed Gamble on 3/19/19.
//  Copyright © 2019 Breadwinner AG. All rights reserved.
//
//  See the LICENSE file at the project root for license information.
//  See the CONTRIBUTORS file at the project root for a list of contributors.

#include "WKUnit.h"
#include "support/BRArray.h"

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <assert.h>

struct WKUnitRecord {
    WKCurrency currency;
    char *uids;
    char *name;
    char *symbol;
    WKUnit base;
    uint8_t decimals;
    WKRef ref;
};

IMPLEMENT_WK_GIVE_TAKE (WKUnit, wkUnit)

static WKUnit
wkUnitCreateInternal (WKCurrency currency,
                          const char *code,
                          const char *name,
                          const char *symbol) {
    WKUnit unit = malloc (sizeof (struct WKUnitRecord));

    unit->currency = wkCurrencyTake (currency);
    unit->name   = strdup (name);
    unit->symbol = strdup (symbol);
    unit->uids   = malloc (strlen (wkCurrencyGetUids(currency)) + 1 + strlen(code) + 1);
    sprintf (unit->uids, "%s:%s", wkCurrencyGetUids(currency), code);

    unit->ref = WK_REF_ASSIGN (wkUnitRelease);
    return unit;
}

extern WKUnit
wkUnitCreateAsBase (WKCurrency currency,
                        const char *code,
                        const char *name,
                        const char *symbol) {
    WKUnit unit = wkUnitCreateInternal (currency, code, name, symbol);

    unit->base = NULL;
    unit->decimals = 0;

    return unit;
}

extern WKUnit
wkUnitCreate (WKCurrency currency,
                  const char *code,
                  const char *name,
                  const char *symbol,
                  WKUnit baseUnit,
                  uint8_t powerOffset) {
    assert (NULL != baseUnit);
    WKUnit unit = wkUnitCreateInternal (currency, code, name, symbol);

    unit->base = wkUnitTake (baseUnit);
    unit->decimals = powerOffset;

    return unit;
}

static void
wkUnitRelease (WKUnit unit) {
    if (NULL != unit->base) wkUnitGive (unit->base);
    wkCurrencyGive (unit->currency);
    free (unit->uids);
    free (unit->name);
    free (unit->symbol);

    memset (unit, 0, sizeof(*unit));
    free (unit);
}

private_extern BRArrayOf(WKUnit)
wkUnitTakeAll (BRArrayOf(WKUnit) units) {
    if (NULL != units)
        for (size_t index = 0; index < array_count (units); index++)
            wkUnitTake(units[index]);
    return units;
}

private_extern BRArrayOf(WKUnit)
wkUnitGiveAll (BRArrayOf(WKUnit) units) {
    for (size_t index = 0; index < array_count (units); index++)
        wkUnitGive(units[index]);
    return units;
}

extern const char *
wkUnitGetUids(WKUnit unit) {
    return unit->uids;
}

extern const char *
wkUnitGetName (WKUnit unit) {
    return unit->name;
}

extern const char *
wkUnitGetSymbol (WKUnit unit) {
    return unit->symbol;
}

extern WKCurrency
wkUnitGetCurrency (WKUnit unit) {
    return wkCurrencyTake (unit->currency);
}

extern WKBoolean
wkUnitHasCurrency (WKUnit unit,
                       WKCurrency currency) {
    return AS_WK_BOOLEAN (unit->currency == currency);
}
extern WKUnit
wkUnitGetBaseUnit (WKUnit unit) {
    return wkUnitTake (NULL == unit->base ? unit : unit->base);
}

extern uint8_t
wkUnitGetBaseDecimalOffset (WKUnit unit) {
    return NULL == unit->base ? 0 : unit->decimals;
}

extern WKBoolean
wkUnitIsCompatible (WKUnit u1,
                        WKUnit u2) {
    return wkCurrencyIsIdentical (u1->currency, u2->currency);
}

extern WKBoolean
wkUnitIsIdentical (WKUnit u1,
                       WKUnit u2) {
    return AS_WK_BOOLEAN (u1 == u2
                              || u1->uids == u2->uids
                              || 0 == strcmp (u1->uids, u2->uids));
}

File: WalletKitJava/WalletKitNative/src/main/java/com/blockset/walletkit/nativex/WKUnit.java
/*
 * Created by Michael Carrara <michael.carrara@breadwallet.com> on 7/1/19.
 * Copyright (c) 2019 Breadwinner AG.  All right reserved.
*
 * See the LICENSE file at the project root for license information.
 * See the CONTRIBUTORS file at the project root for a list of contributors.
 */
package com.blockset.walletkit.nativex;

import com.blockset.walletkit.nativex.library.WKNativeLibraryDirect;
import com.google.common.primitives.UnsignedBytes;
import com.google.common.primitives.UnsignedInteger;
import com.sun.jna.Pointer;
import com.sun.jna.PointerType;

public class WKUnit extends PointerType {

    public static WKUnit createAsBase(WKCurrency currency, String uids, String name, String symbol) {
        return new WKUnit(
                WKNativeLibraryDirect.wkUnitCreateAsBase(
                        currency.getPointer(),
                        uids,
                        name,
                        symbol
                )
        );
    }

    public static WKUnit create(WKCurrency currency, String uids, String name, String symbol, WKUnit base, UnsignedInteger decimals) {
        byte decimalsAsByte = UnsignedBytes.checkedCast(decimals.longValue());
        return new WKUnit(
                WKNativeLibraryDirect.wkUnitCreate(
                        currency.getPointer(),
                        uids,
                        name,
                        symbol,
                        base.getPointer(),
                        decimalsAsByte
                )
        );
    }

    public WKUnit() {
        super();
    }

    public WKUnit(Pointer address) {
        super(address);
    }

    public String getUids() {
        Pointer thisPtr = this.getPointer();

        return WKNativeLibraryDirect.wkUnitGetUids(thisPtr).getString(0, "UTF-8");
    }

    public String getName() {
        Pointer thisPtr = this.getPointer();

        return WKNativeLibraryDirect.wkUnitGetName(thisPtr).getString(0, "UTF-8");
    }

    public String getSymbol() {
        Pointer thisPtr = this.getPointer();

        return WKNativeLibraryDirect.wkUnitGetSymbol(thisPtr).getString(0, "UTF-8");
    }

    public UnsignedInteger getDecimals() {
        Pointer thisPtr = this.getPointer();

        return UnsignedInteger.fromIntBits(UnsignedBytes.toInt(WKNativeLibraryDirect.wkUnitGetBaseDecimalOffset(thisPtr)));
    }

    public WKUnit getBaseUnit() {
        Pointer thisPtr = this.getPointer();

        return new WKUnit(WKNativeLibraryDirect.wkUnitGetBaseUnit(thisPtr));
    }

    public WKCurrency getCurrency() {
        Pointer thisPtr = this.getPointer();

        return new WKCurrency(WKNativeLibraryDirect.wkUnitGetCurrency(thisPtr));
    }

    public boolean hasCurrency(WKCurrency currency) {
        Pointer thisPtr = this.getPointer();

        return WKBoolean.WK_TRUE == WKNativeLibraryDirect.wkUnitHasCurrency(thisPtr,  currency.getPointer());
    }

    public boolean isCompatible(WKUnit other) {
        Pointer thisPtr = this.getPointer();

        return WKBoolean.WK_TRUE == WKNativeLibraryDirect.wkUnitIsCompatible(thisPtr, other.getPointer());
    }

    public boolean isIdentical(WKUnit other) {
        Pointer thisPtr = this.getPointer();

        return WKBoolean.WK_TRUE == WKNativeLibraryDirect.wkUnitIsIdentical(thisPtr, other.getPointer());
    }

    public void give() {
        Pointer thisPtr = this.getPointer();

        WKNativeLibraryDirect.wkUnitGive(thisPtr);
    }
}
